src/univers/version_constraint.py

```diff
@@ -0,0 +1,221 @@
+        >>> v22 = PypiVersion("2.2")
+        >>> v23 = PypiVersion("2.3")
+        >>> v24 = PypiVersion("2.4")
+        >>> assert v23 in VersionConstraint(comparator="=", version=v23)
+        >>> assert v24 not in VersionConstraint(comparator="=", version=v23)
+        >>> assert None not in VersionConstraint(comparator="=", version=v23)
+
+        >>> assert None not in VersionConstraint(comparator="!=", version=v24)
+        >>> assert v22 in VersionConstraint(comparator="!=", version=v23)
+        >>> assert v23 in VersionConstraint(comparator="!=", version=v24)
+        >>> assert v24 not in VersionConstraint(comparator="!=", version=v24)
+
+        >>> assert v24 in VersionConstraint(comparator=">", version=v23)
+        >>> assert v23 not in VersionConstraint(comparator=">", version=v23)
+        >>> assert v24 in VersionConstraint(comparator=">=", version=v23)
+        >>> assert v23 in VersionConstraint(comparator=">=", version=v23)
+        >>> assert v22 not in VersionConstraint(comparator=">=", version=v23)
+
+        >>> assert v23 in VersionConstraint(comparator="<", version=v24)
+        >>> assert v23 in VersionConstraint(comparator="<=", version=v24)
+        >>> assert v24 in VersionConstraint(comparator="<=", version=v24)
+        >>> assert v24 not in VersionConstraint(comparator="<", version=v24)
+        """
+        if not version:
+            return False
+
+        # TODO: consider if we should accept any type and return False?
+        if version.__class__ != self.version.__class__:
+            raise ValueError(
+                f"Cannot compare {version.__class__!r} instance "
+                f"with {self.version.__class__!r} instance."
+            )
+        try:
+            comp_operator = COMPARATORS[self.comparator]
+        except KeyError as e:
+            raise ValueError(f"Unknown comparator: {self.comparator}") from e
+
+        return comp_operator(self.version, version)
+
+    contains = __contains__
+
+    @classmethod
+    def validate(cls, constraints):
+        """
+        Raise an assertion error if the ``constraints`` is not a two-level
+        nested list of VersionConstraint objects.
+        """
+        assert isinstance(constraints, (list, tuple)), constraints
+        for inner_constraints in constraints:
+            assert isinstance(inner_constraints, (list, tuple)), inner_constraints
+            for constraint in inner_constraints:
+                assert isinstance(constraint, VersionConstraint), constraint
+
+    @classmethod
+    def sort(cls, constraints):
+        """
+        Return sorted nested list of ``constraints`` using the "vers" canonical
+        order. Sorting is done in place.
+        """
+        for inner_constraints in constraints:
+            inner_constraints.sort(key=lambda vc: str(vc))
+        constraints.sort(key=lambda vc: str(vc))
+        return constraints
+
+    @classmethod
+    def to_constraints_string(cls, constraints):
```

Comment:
Are doctest possible for this ? Maybe add some examples.

Reply:
Agreed